### PR TITLE
Make `typeof` start a nested lexical scope

### LINKED ISSUE #10521: `typeof` should create a new scope
Variables declared and assigned inside a `typeof` expression have no effect on code outside, so they shouldn't be accessible from outside at all. The following error:

```crystal
x = typeof(y = 1)
y # Error: read before assignment to local variable 'y'
```

should have been `undefined local variable or method 'y' for top-level`. In particular, the variables shouldn't be accessible from _other_ `typeof` expressions, like:

```crystal
module A; end
module B; end

typeof(begin
  x = uninitialized A
  x.is_a?(B) ? x : raise "" # #2404
end)
typeof(begin
  x = uninitialized B # Error: variable 'x' already declared with type A
  x.is_a?(A) ? x : raise ""
end)
```

### PR DESCRIPTION
**Pull request by HertzDevil**
Fixes #10521. The inner expressions do not each reset the variable scope.

The first commit contains the entire fix. The second commit further refactors the parser so that it becomes clearer which and when variable scopes are created by the parser.

### CODE REVIEW

File: src/compiler/crystal/syntax/parser.cr [hunk 1]
```diff
@@ -5976,23 +5968,28 @@ module Crystal
...
-      push_def(Set.new(args.map &.name))
-      ret = yield
-      pop_def
-      ret
+      begin
+        @var_scopes.push(Set(String).new)
+        yield
+      ensure
+        @var_scopes.pop
+      end
```
Comment (Sija) [inline src/compiler/crystal/syntax/parser.cr:5982]:
```suggestion
      @var_scopes.push(Set(String).new)
      yield
    ensure
      @var_scopes.pop
```

**Review by Sija — COMMENTED** [review]
One small style comment.

File: src/compiler/crystal/syntax/parser.cr [hunk 2]
```diff
@@ -5976,23 +5968,28 @@ module Crystal
...
     end
 
-    def push_def(set)
-      @def_vars.push(set)
+    # Creates a new variable scope with the same variables as the current scope,
+    # and then returns the yield result, resettin the scope afterwards.
```
Comment (straight-shoota) [inline src/compiler/crystal/syntax/parser.cr:5986]:
```suggestion
    # and then returns the yield result, resetting the scope afterwards.
```

**Review by straight-shoota — APPROVED** [review]

File: src/compiler/crystal/syntax/parser.cr (hunk 1 above)
Comment (straight-shoota) [reply src/compiler/crystal/syntax/parser.cr:5982]: That would incorrectly pop for the early return branch.

**Comment by straight-shoota** [comment]
That CI error looks... interesting ;)

**Comment by HertzDevil** [comment]
It seems `@meta_vars`'s identity cannot be clobbered. Hoping this workaround passes...

**Review by Hayato5648 — APPROVED** [review]

**Review by beta-ziliani — APPROVED** [review]
Besides the fix itself, I really like the refactor done in the code. Thanks @HertzDevil !

**Pull request merged**